### DISCUSSION TITLE
Help in Deploying Django, Django Channels to AWS

### DISCUSSION THREAD

**Discussion by PaulleDemon** [PaulleDemon/AWS-deployment#7] [Help with deployment]
If you require any help with deploying Django, and Django channels to AWS, tag me in this thread.  (note: standard charges apply).

_You must also be able to provide the necessary IAM permissions for me to help you. If you are unable to do so, due to company policy or other reasons, I  suggest you buy this video and do it yourself: https://www.buymeacoffee.com/artpaul/e/241689, you would still receive my support, if you require._

**Comment by HammadAhm3d** [comment]
@PaulleDemon Thank you for this guide. I have been following this to serve notifications from my django application via a websocket connection. I have followed every step and also passed the {"foo": "bar"} test that you mentioned at the end of the file. However, when I try to connect to the websocket using the elastic beanstalk link, it fails. I looked at the logs and the logs show that the endpoint ws/ (or the endpoint for websocket) is not found (404). This works totally fine on the localhost though. I can post the error logs if you require.

**Reply by PaulleDemon** [reply]
Hey can you post the logs. 

also did you follow this part: https://github.com/PaulleDemon/AWS-deployment/blob/master/django-channels.md ?

**Reply by HammadAhm3d** [reply]
Yes I did follow it. Here is the eb-engine.log file:
[eb-engine.log](https://github.com/user-attachments/files/17488762/eb-engine.log)

If you notice in the logs, at 404 error is thrown at the websocket endpoint i.e, ws/notifications

**Reply by PaulleDemon** [reply]
show me the log where the error occurs. Also, did you ssh into your instance and try the foo bar test as mentioned [here](https://github.com/PaulleDemon/AWS-deployment/blob/master/django-channels.md#debugging-tips)?

**Reply by HammadAhm3d** [reply]
Yes I did and it worked! Here's the error block:

Oct 23 09:49:01 ip-172-31-43-210 web[9645]: During handling of the above exception, another exception occurred:
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Traceback (most recent call last):
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:  File "/var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/django/template/base.py", line 883, in _resolve_lookup
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:    current = current[bit]
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:              ~~~~~~~^^^^^
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: TypeError: 'URLResolver' object is not subscriptable
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: During handling of the above exception, another exception occurred:
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Traceback (most recent call last):
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:  File "/var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/django/template/base.py", line 893, in _resolve_lookup
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:    current = getattr(current, bit)
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:              ^^^^^^^^^^^^^^^^^^^^^
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: AttributeError: 'URLResolver' object has no attribute 'name'
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: During handling of the above exception, another exception occurred:
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Traceback (most recent call last):
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:  File "/var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/django/template/base.py", line 899, in _resolve_lookup
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:    current = current[int(bit)]
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:                      ^^^^^^^^
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: ValueError: invalid literal for int() with base 10: 'name'
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: During handling of the above exception, another exception occurred:
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Traceback (most recent call last):
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:  File "/var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/django/template/base.py", line 906, in _resolve_lookup
Oct 23 09:49:01 ip-172-31-43-210 web[9645]:    raise VariableDoesNotExist(
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: django.template.base.VariableDoesNotExist: Failed lookup for key [name] in <URLResolver <URLPattern list> (admin:admin) 'admin/'>
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Not Found: /ws/notifications/
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Not Found: /ws/notifications/

I have routing.py
`from django.urls import re_path
from core.Notification.consumers import NotificationConsumer

websocket_urlpatterns = [
    re_path(r'ws/notifications/$', NotificationConsumer.as_asgi()),
]
`

asgi.py:
`import os
import django
from django.core.asgi import get_asgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'project_management.settings')

django.setup()

from channels.auth import AuthMiddlewareStack
from channels.routing import ProtocolTypeRouter, URLRouter
from project_management.routing import websocket_urlpatterns

application = ProtocolTypeRouter({
    "http": get_asgi_application(),
    "websocket": AuthMiddlewareStack(
        URLRouter(
            websocket_urlpatterns
        )
    ),
})
`

channels in settings.py:
`CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG': {
            "hosts": [('XXXX-redis-endpoint-XXXX.amazonaws.com', 6379)],
        },
    },`

**Reply by HammadAhm3d** [reply]
@PaulleDemon Any news?

**Reply by PaulleDemon** [reply]
@HammadAhm3d sorry was busy, this has nothing to do with aws, you need to check your model code to see if your returning the correct value, 

```
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: current = current[int(bit)]
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: ^^^^^^^^
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: ValueError: invalid literal for int() with base 10: 'name'
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: During handling of the above exception, another exception occurred:
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: Traceback (most recent call last):
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: File "/var/app/venv/staging-LQM1lest/lib64/python3.11/site-packages/django/template/base.py", line 906, in _resolve_lookup
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: raise VariableDoesNotExist(
Oct 23 09:49:01 ip-172-31-43-210 web[9645]: django.template.base.VariableDoesNotExist: Failed lookup for key [name] in <URLResolver (admin:admin) 'admin/'>
```
